libs/pilight/config/settings.c

```diff
@@ -458,6 +458,90 @@ static int settings_parse(JsonNode *root) {
 			} else {
 				settings_add_string(jsettings->key, jsettings->string_);
 			}
+#ifdef EVENTS
+		} else if(strcmp(jsettings->key, "smtp-sender") == 0 ||
+					strcmp(jsettings->key, "smtp-user") == 0) {
+			if(jsettings->tag != JSON_STRING) {
+				logprintf(LOG_ERR, "config setting \"%s\" must contain an e-mail address", jsettings->key);
+				have_error = 1;
+				goto clear;
+			} else if(jsettings->string_ == NULL) {
+				logprintf(LOG_ERR, "config setting \"%s\" must contain an e-mail address", jsettings->key);
+				have_error = 1;
+				goto clear;
+			} else if(strlen(jsettings->string_) > 0) {
+#if !defined(__FreeBSD__) && !defined(_WIN32)
+				char validate[] = "^[a-zA-Z0-9_.]+@([a-zA-Z0-9]+\\.)+[a-zA-Z0-9]{2,3}$";
+				reti = regcomp(&regex, validate, REG_EXTENDED);
+				if(reti) {
+					logprintf(LOG_ERR, "could not compile regex for %s", jsettings->key);
+					have_error = 1;
+					goto clear;
+				}
+				reti = regexec(&regex, jsettings->string_, 0, NULL, 0);
+				if(reti == REG_NOMATCH || reti != 0) {
+					logprintf(LOG_ERR, "config setting \"%s\" must contain an e-mail address", jsettings->key);
+					have_error = 1;
+					regfree(&regex);
+					goto clear;
+				}
+#endif
+			settings_add_string(jsettings->key, jsettings->string_);
+			}
+		} else if(strcmp(jsettings->key, "smtp-password") == 0) {
+			if(jsettings->tag != JSON_STRING) {
+				logprintf(LOG_ERR, "config setting \"%s\" must contain a password string", jsettings->key);
+				have_error = 1;
+				goto clear;
+			} else if(jsettings->string_ == NULL) {
+				logprintf(LOG_ERR, "config setting \"%s\" must contain a password string", jsettings->key);
+				have_error = 1;
+				goto clear;
+			} else {
+				settings_add_string(jsettings->key, jsettings->string_);
+			}
+		} else if(strcmp(jsettings->key, "smtp-host") == 0) {
+			if(jsettings->tag != JSON_STRING) {
+				logprintf(LOG_ERR, "config setting \"%s\" must contain an smtp host address in the format of \"smtp.gmail.com\"", jsettings->key);
+				have_error = 1;
+				goto clear;
+			} else if(jsettings->string_ == NULL) {
+				logprintf(LOG_ERR, "config setting \"%s\" must contain an smtp host address in the format of \"smtp.gmail.com\"", jsettings->key);
+				have_error = 1;
+				goto clear;
+			} else if(strlen(jsettings->string_) > 0) {
+#if !defined(__FreeBSD__) && !defined(_WIN32)
+				char validate[] = "^(smtp)\\.[a-zA-Z0-9_]+\\.[a-zA-Z]{2,3}$";
```

Comment:
Did some searching for smtp server adresses and came to the conclusion 
that the current regex is much to strict.

Examples:

smtp.x.y
mail.x.y
mailhost.x.y
relay.x.y
post.x.y
mail.username.x.y
mail.domainname.x.y

These are just Dutch smtp hosts, so there is a big chance that there 
will be more.
So, if you agree,  I will change the regex to:

^([a-zA-Z0-9\\_\\-]){2,20}(\\.([a-zA-Z0-9\\_\\-]){2,20}){2,3}$

This will allow any prefix ,subdomains, and TLD's with more than 3 
characters (eg .amsterdam) and will still catch the most common typo's.

CurlyMoo schreef op 29-3-2015 om 12:45:
>
> In libs/pilight/config/settings.c 
> <https://github.com/pilight/pilight/pull/163#discussion_r27353348>:
>
> > +				goto clear;
> > +			} else {
> > +				settings_add_string(jsettings->key, jsettings->string_);
> > +			}
> > +		} else if(strcmp(jsettings->key, "smtp-host") == 0) {
> > +			if(jsettings->tag != JSON_STRING) {
> > +				logprintf(LOG_ERR, "config setting \"%s\" must contain an smtp host address in the format of \"smtp.gmail.com\"", jsettings->key);
> > +				have_error = 1;
> > +				goto clear;
> > +			} else if(jsettings->string_ == NULL) {
> > +				logprintf(LOG_ERR, "config setting \"%s\" must contain an smtp host address in the format of \"smtp.gmail.com\"", jsettings->key);
> > +				have_error = 1;
> > +				goto clear;
> > +			} else if(strlen(jsettings->string_) > 0) {
> > +#if !defined(__FreeBSD__) && !defined(_WIN32)
> > +				char validate[] = "^(smtp)\\.[a-zA-Z0-9_]+\\.[a-zA-Z]{2,3}$";
>
> I'm not sure this is a valid regex: |smtp.blabla.co.uk|.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/pilight/pilight/pull/163/files#r27353348>.
>